docs/proposals/20210222-kubelet-authentication.md

```diff
@@ -0,0 +1,319 @@
+
+### Risks and Mitigations
+
+There may be additional security risks being introduced in this design. In order to mitigate this, this proposal will be taken to SIG Security and SIG Auth for review **before the beta graduation**.
+
+## Alternatives
+
+#### Implement within the cloud providers instead of Cluster API
+Given that there is an extant implementation in cloud-provider-gcp, this could be extended to all of the cloud providers. However, there are some advantages to making Cluster API responsible for kubelet registration:
+- Cluster API can create the node resource as part of the CSR approval process and apply the relevant taints, labels and roles up front. The Kubernetes Cloud Provider contract has no understanding of the desired state of a node.
+- No changes to the assumptions around connectivity between management and workload clusters are required, neither does the signer need to be included as a static pod during control plane instantiation.
+
+#### Implement as authentication webhook, as per aws-iam-authenticator (Amazon EKS)
+If attestation was implemented as an authentication webhook, it would be in the critical path for all token-based authentication against the API server. It would also additionally be needed to be set up at workload cluster instantiation via a static pod and API server start up.
+
+
+## Upgrade Strategy
+
+for the upgrade strategy please refer to the cluster API node bootstrapper proposal
+
+## Additional Details
+
+### Test Plan [optional]
+
+- E2E tests to be added to AWS, vSphere and Azure providers as each provider implements the signer
+- E2E tests to be added for the insecure signers for use with CAPD.
+Upgrade tests from latest minor release to latest main branch of Kubernetes 
+
+
+### Graduation Criteria [optional]
+
+#### Graduation to beta
+- E2E tests testing upgrades to latest main branch of Kubernetes are required such that Cluster API can make appropriate changes to node registration if kubelet or kubeadm behaviour changes.
+- Security review by SIG Auth and SIG Security
+#### Graduation to GA
+- Beta for 2 releases
+- Optionally, external security review of the combined Cluster API and kubeadm model.
+
+
+### Version Skew Strategy [optional]
+
+There shouldn’t be any version skew issues between nodeadm  and the bootstrapper, As the only interaction between the two is through the CSR API.
+Any changes to the attestation data should be handled in a backward compatible manner by the infrastructure provider when implementing the interface used by the node-attestation-controller, by making sure it's able to convert an older attestation format to a newer one.
+
+
+## Implementation History
+
+- [ ] MM/DD/YYYY: Proposed idea in an issue or [community meeting]
+- [ ] MM/DD/YYYY: Compile a Google Doc following the CAEP template (link here)
+- [ ] MM/DD/YYYY: First round of feedback from community
+- [ ] MM/DD/YYYY: Present proposal at a [community meeting]
+- [ ] MM/DD/YYYY: Open proposal PR
```

Comment:
were some of the above steps skipped?

Reply:
I can add the google doc for the new bootstrapper proposal,  @randomvariable and I wanted split node attestation into its own proposal for a better reviewing experience

Reply:
Updated